Add default "resolvedIf" function for tasks

binokaryg opened the pull request:
# Description

Adds a default `resolvedIf` function for tasks.

medic/medic-conf#107

## defaultResolvedIf
- Uses [Utils.isFormSubmittedInWindow](https://github.com/medic/medic-nootils/blob/4339b25201bfbf51eef9179c8b7e0e92734a0d48/src/nootils.js#L57) to check if the resolving form has been submitted within a time period:
    - For contact-based tasks, the period is the same as the task window period
    - For report based tasks, the period is determined as follows:
        - start: start of the task window or 1 millisecond after reported date of the source report, whichever comes later
        - end: end of the task window
    - The resolving form is the task's first action form


## Usage
Options:
A. Skip `resolvedIf` section in the task definition, `defaultResolvedIf` will be used automatically
B. Using `this.defaultResolvedIf`:
1. Without specifying the parameters:
    ```
    resolvedIf: this.defaultResolvedIf,
    ```

2. Specifying parameters and optionally add other conditions:
    ```
    resolvedIf: function (contact, report, event, dueDate) {
      return this.defaultResolvedIf(contact, report, event, dueDate) && ...
    },
    ```
3. Specifying the resolving form type if it is different from the default (form type specified in the first action of the task):
    ```
    resolvedIf: function (contact, report, event, dueDate) {
      return this.defaultResolvedIf(contact, report, event, dueDate, 'delivery');
    },
    ```
# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or integration tests where appropriate
- Backwards compatible: Works with existing data and configuration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

binokaryg committed:
Adds a default resolvedIf condition which is supported in 3 ways:
1. Omitting `resolvedIf` in the task definition: It uses `Utils.defaultResolvedIf`. (from medic-nootils)
2. Utils.defaultResolvedIf(c, r, event, dueDate, resolvingForm). To be added to medic-nootils. All but last arguments are the same as of `resolvedIf()`. The last argument is `resolvingForm`.
3. this.resolvedIf(c, r, event, dueDate).  resolvingForm can also be passed but it is optional. If not available, it will take from `this.definition.actions[0].form`. Uses `Utils.defaultResolvedIf`.

binokaryg committed:
Adding some tests for medic/medic-conf#107

kennsippell commented on test/run-nools-lib.js:
```diff
@@ -19,6 +19,28 @@ const runNoolsLib = ({ c, targets, tasks }) => {
+        });
+        return result;
+      },
+      defaultResolvedIf: function (contact, report, event, dueDate, resolvingForm) {
```
weird that this is duplicated in test code...

binokaryg replied:
removed

kennsippell commented on test/nools/task-emitter.spec.js:
```diff
@@ -287,6 +306,25 @@ describe('task-emitter', () => {
         });
       });
 
+      it('resolvedIf is not required', () => {
```
shouldn't have two tests with the same name. what's the difference?

binokaryg replied:
One was under "person based task" and another under "report based task", but that might be unnecessary, so only keeping one

kennsippell reviewed (approved):
u get it

binokaryg committed:
#107
using Utils.isFormSubmittedInWindow instead of Utils.defaultResolvedIf
more tests

binokaryg committed:
#107 define defaultResolvedIf in single place, one more test
